### DISCUSSION TITLE
【主库】Block access and wait for recovery: xxx

### DISCUSSION THREAD

**Discussion by 2881099** [dotnetcore/FreeSql#1080] [Q&A]
### 【主库】Block access and wait for recovery: xxx

---

一般是数据库连接失败，才会出现，请检查程序与数据库之间的网络。具体按 xxx 给出的提示进行排查。

1、FreeSql UseConnectionString 的机制是，一旦发生服务器错误，则会将整个对象不可用，可以把它理解为断熔。之后 3-5 秒所有 IFreeSql 操作都会直接抛出异常：【主库】Block access and wait for recovery: xxx，不会继续冲击数据库服务器。

2、FreeSql UseConnectionFactory 的机制是，跟传统 ado.net 一样，不会发生该错误，使用方法如下：

```c#
 var fsql = new FreeSqlBuilder()
    .UseConnectionFactory(DataType.SqlServer, () => new SqlConnection(...))
    .Build();
 services.AddSingleton(fsql);
```

**Comment by 2881099** [comment]
2.6.100升级到3.x 后无法连接 sqlserver 提示证书无效, 提示证书链是由不受信任的颁发机构颁发的. 
请尝试: 

1.连接字符串里加入 “Encrypt=True; TrustServerCertificate=True;” 

2.使用FreeSql.Provider.SqlServerForSystem替换FreeSql.Provider.SqlServer 

深入讨论请转到 https://github.com/dotnetcore/FreeSql/issues/992#issuecomment-1005305027

**Reply by coco-bike** [reply]
如果是mysql发生这样的事怎么解决？运行一段时间就会报错：
【主库】Block access and wait for recovery: Unable to connect to any of the specified MySQL hosts

**Reply by 2881099** [reply]
百度：Unable to connect to any of the specified MySQL hosts

**Comment by xbin-y** [comment]
> ### 【主库】状态不可用，等待后台检查程序恢复方可使用。xxx
> ### 【主库】Block access and wait for recovery: xxx
> 一般是数据库连接失败，才会出现，请检查程序与数据库之间的网络。具体按 xxx 给出的提示进行排查。
> 
> 1、FreeSql UseConnectionString 的机制是，一旦发生服务器错误，则会将整个对象不可用，可以把它理解为断熔。之后 3-5 秒所有 IFreeSql 操作都会直接抛出异常：【主库】Block access and wait for recovery: xxx，不会继续冲击数据库服务器。
> 
> 2、FreeSql UseConnectionFactory 的机制是，跟传统 ado.net 一样，不会发生该错误，使用方法如下：
> 
> ```cs
>  var fsql = new FreeSqlBuilder()
>     .UseConnectionFactory(DataType.SqlServer, () => new SqlConnection(...))
>     .Build();
>  services.AddSingleton(fsql);
> ```

使用UseConnectionFactory时，当 AsTreeCte() 无法知道正确的serverVersion，生成的sql会有语法错误

**Reply by 2881099** [reply]
```c#
using (var conn = new MySqlConnection(string))
{
   var list = fsql.Select<T>().WithConnection(conn).AsTreeCte(...).ToList();
}
```
试试这样。

**Comment by xiaoxiong3533521** [comment]
这个坑应该写到文档里面,    我也是遇到这个问题好久,没有找到解决办法,

**Comment by xiaoxiong3533521** [comment]
我现在遇到的问题是  除了这个错误,同时还会出现另外一个错误:  通常每个套接字地址(协议/网络地址/端口) 只运行使用一次. 127.0.0.1:3306
